### PR TITLE
Support for electron-main process customization with DI

### PR DESCRIPTION
**Pull request by kittaakos**
#### What it does

This PR consists of the
 - original #7964 PR, and
 - #8069 PR, that fixes a few macOS issues and the tests.

Closes #7964 
Closes #8069 

This PR contributes a JSON-RPC communication scheme between the `electron-main` process and the `renderer` processes. It is reusing the same architecture as the websocket-based connection between frontend/backend, but adapted to work over electron's ipc apis. Currently there isn't scoped connections, meaning that each renderer process will interact with the same singleton in the electron-main process, but for the current use-cases this is sufficient.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by marechal-p** [comment]
Thank you for the changes! I fixed nitpicks I found, but I'm happy with this PR.

Although I'm having trouble marking the `Added dummy [...]` commit as authored by yourself... Can you maybe do `git commit --amend --reset-author` and force push again? Sorry for the inconvenience. I remember I correctly fixed commit authorship before, but cannot now for some reason.

**Comment by kittaakos** [comment]
> Can you maybe do `git commit --amend --reset-author` and force push again? Sorry for the inconvenience.

I am fine with it, I do not really mind who's avatar is next to the commit. Is it a problem for you?

**Comment by akosyakov** [comment]
Could you elaborate please in `What it does`: how different parts to talk to each other and how JSON-RPC multiplexing works in context of Electron, just to understand?

 `common` means that electron APIs can be used, what `electron-common` means? Why it cannot just go to `common`?

File: .travis.yml [hunk 1]
```diff
@@ -15,6 +15,7 @@ cache:
     - dev-packages/cli/node_modules
     - examples/api-samples/node_modules
     - examples/browser/node_modules
+    - examples/electron-api-samples/node_modules
```
Comment (akosyakov) [inline .travis.yml:18]: do we really need a new package? Could we add the second extension entry point in `api-samples` instead?

File: packages/electron/src/electron-main/electron-application.ts [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+     */
+    requestStop(): void {
+        app.quit();
+    }
+
+    protected async handleMainCommand(params: ExecutionParams, options: MainCommandOptions): Promise<void> {
+        if (typeof options.file === 'undefined') {
+            await this.openDefaultWindow();
+        } else {
+            await this.openWindowWithWorkspace(realpathSync(path.resolve(params.cwd, options.file)));
```
Comment (marechal-p) [inline packages/electron/src/electron-main/electron-application.ts:214]: `realpathSync` can fail here when `options.file` doesn't exists, this is my bad.

Tell me if you want me to fix this, or if you'll do it. I don't want to step on your toes.

**Comment by kittaakos** [comment]
> I fixed nitpicks I found,

Alright, what exactly? It would be nice to share this info with others. I thought you will review my part instead of modifying and force pushing it; so we can see the changes.

File: packages/electron/src/electron-main/electron-application.ts (hunk 2 above)
Comment (kittaakos) [reply packages/electron/src/electron-main/electron-application.ts:214]: Go ahead, please. It's your PR eventually, I just wanted to verify the functionality by adding a sample.

File: packages/electron/src/electron-main/electron-application.ts [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+    protected config: FrontendApplicationConfig;
+    readonly backendPort = new Deferred<number>();
+
+    async start(config: FrontendApplicationConfig): Promise<void> {
+        this.config = config;
+        this.hookApplicationEvents();
+        const port = await this.startBackend();
+        this.backendPort.resolve(port);
+        await app.whenReady();
+        await this.attachElectronSecurityToken(port);
```
Comment (marechal-p) [inline packages/electron/src/electron-main/electron-application.ts:156]: @kittaakos I replaced the await on the deferred promise you had before.

**Comment by akosyakov** [comment]
Let's merge after the release, that's an arch change to merge in 3 days before the release.

File: packages/electron/src/electron-main/electron-application.ts [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+    protected onWillQuit(event: ElectronEvent): void {
+        this.stopContributions();
+    }
+
+    protected async onSecondInstance(event: ElectronEvent, argv: string[], cwd: string): Promise<void> {
+        await this.launch({ argv, cwd, secondInstance: true });
+    }
+
+    protected onWindowAllClosed(event: ElectronEvent): void {
+        this.requestStop();
```
Comment (marechal-p) [inline packages/electron/src/electron-main/electron-application.ts:420]: @kittaakos I aligned the handling of the event with the others handlers so that it's easier to extend.

File: packages/electron/src/electron-main/electron-application.ts (hunk 2 above)
Comment (kittaakos) [reply packages/electron/src/electron-main/electron-application.ts:214]: >It's your PR eventually

I hope you have taken it as a compliment ;)

Comment (marechal-p) [reply packages/electron/src/electron-main/electron-application.ts:214]: Ok I'll force push again then.

File: .travis.yml (hunk 1 above)
Comment (kittaakos) [reply .travis.yml:18]: Sure, why is it a problem having a dedicated sample for electron?

Comment (kittaakos) [reply .travis.yml:18]: @marechal-p, please ping me when you're done, and I move the sample.

Comment (marechal-p) [reply .travis.yml:18]: @kittaakos it is all yours.

**Comment by marechal-p** [comment]
> looking at sample looks quite clean :+1 adding a short PR description about new architecture would be helpful in the future

You are entirely right, I added a simple description in [`packages/electron/README.md`](https://github.com/eclipse-theia/theia/blob/f875177139ae09bfceafc222ebae57045dca67e8/packages/electron/README.md).

I also updated the PR description.

File: .travis.yml (hunk 1 above)
Comment (kittaakos) [reply .travis.yml:18]: I am not sure if we want to move it to the existing sample project because if we move it to there, the browser example will depend on it, which pulls the electron dependency in. Let me check, if it works, I will move it. Otherwise, I'll leave here as originally wanted.

**Comment by marechal-p** [comment]
> `common` means that electron APIs can be used, what `electron-common` means? Why it cannot just go to `common`?

It is interfaces really specific to electron. Someone importing those interfaces from a `browser` script should really feel bad seeing something coming from `electron-common`, hopefully avoiding small mistakes.

**Comment by marcdumais-work** [comment]
> the nitpicks I had were so insignificant that I didn't want to just ask when I could just fix it. Sorry if you don't like me pushing to your branch, I will not do it again.

I just wanted to mention that GH has a feature that can help for such little "nitpicks", in a more transparent way: one can suggest changes, that the PR author can accept at the click of a button. See:  https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request#applying-a-suggested-change

File: packages/electron/package.json [hunk 5]
```diff
@@ -14,12 +14,22 @@
...
+    "src",
+    "*.js"
+  ],
   "bin": {
     "electron": "electron-cli.js",
     "electron-codecs-test": "electron-codecs-test.js",
     "electron-replace-ffmpeg": "electron-replace-ffmpeg.js"
   },
+  "devDependencies": {
+    "@theia/cli": "^1.2.0"
```
Comment (akosyakov) [inline packages/electron/package.json:29]: why do we need it here?

**Comment by akosyakov** [comment]
Why did we decide to pull code in `@theia/electron`? How do you envision working on the same product supporting browser and electron without adding an explicit dependency to `@theia/electron`?

**Comment by kittaakos** [comment]
> How do you envision working on the same product supporting browser and electron without adding an explicit dependency to `@theia/electron`?

As as a consumer of this change, I had the following understanding:
 - Let assume, you have an extension (`my-ext-main`) that works in both electron and web env, 👍 
 - Your end product's `package.json` contains a bunch of `@theia` dependencies plus `my-ext-main`, we are still good 👍 
 - Then you need something electron specific. You introduce a new `my-ext-electron` extension. This extension explicitly depends on `@theia/electron` and now you have to have a separate `package.json` for another product definition, and this `package.json` contains `my-ext-main`. The other does not.
 - This seems to be a good approach. Did I overlook something?

File: packages/electron/package.json (hunk 5 above)
Comment (kittaakos) [reply packages/electron/package.json:29]: I will check whether we need it or it's a leftover.

**Comment by akosyakov** [comment]
> This seems to be a good approach. Did I overlook something?

The original approach that the same extension can contribute entry point with overrides for Electron. It looks like this PR makes it impossible as well as breaking downstream extensions doing it. It is visible in the PR itself since window entry point was unnecessary moved to `@theia/electron`.

**Comment by kittaakos** [comment]
> It is visible in the PR itself since window entry point was unnecessary moved to `@theia/electron`.

That does not matter, from a downstream project's perspective, it's not a change. I never referenced the electron-specific window implementation, and I had to depend on `@theia/electron` anyway.

**Comment by akosyakov** [comment]
Discussed with @kittaakos offline. It is indeed no-op for down streams, but there is a concern with how we divide code on extensions and js modules:
- the responsibility of an extension should be to contribute some functionality
- within the extension we separate which APIs can be used according to https://github.com/eclipse-theia/theia/wiki/Code-Organization

Introducing new extensions to separate by APIs goes against it. It would be good to revert changes to `@theia/electron` and merge new code into  `@theia/core` and `@theia/api-samples`. If there is no any technical issues preventing it.

**Comment by kittaakos** [comment]
It seems one of the force pushes broke the open new workspace command. No matter what I try to open, it does not open anything and I have the purplish status bar in the browser window. :/ I will fix it as part of the `dev-package` moving.

**Comment by kittaakos** [comment]
OK, I have moved back everything that originally was under `dev-packages/electron`, and moved the new things under `packages/core`. The new things are mainly under `packages/core/src/electron-main`. I also wiped the electron example extension, and moved the dummy updater to `api-samples`. I am aware of at least two issues:
 - when I refresh the browser window, the backend contributions are not disposed. I guess we changed something accidentally, and the WS connection reconnects. It should not.
 - when I start the app from a terminal and have opened at least two windows. I see an error when I terminate the app from the terminal (on macOS) with `Ctrl+C`.

I am looking into these now.

**Comment by kittaakos** [comment]
To discuss:
 - [ ] https://github.com/eclipse-theia/theia/pull/7964/files#r445380379,
 - [ ] (_add your item_)

**Comment by marechal-p** [comment]
Small heads up that I'll look into this again not before the coming monday.

File: packages/core/src/node/cli.ts [hunk 6]
```diff
@@ -25,32 +25,47 @@ export const CliContribution = Symbol('CliContribution');
...
+    ) {
         const pack = require('../../package.json');
-        const version = pack.version;
-        const command = yargs.version(version);
-        command.exitProcess(this.isExit());
-        for (const contrib of this.contributionsProvider.getContributions()) {
-            contrib.configure(command);
-        }
-        const args = command
+        this.parser = yargs
```
Comment (akosyakov) [inline packages/core/src/node/cli.ts:48]: Could you elaborate on changes in this file?

btw you can use `postConstruct` annotation to avoid constructor injection

File: packages/core/src/electron-common/messaging/electron-connection-handler.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+/**
+ * Name of the channel used with `ipcMain.on/emit`.
+ */
+export const THEIA_ELECTRON_IPC_CHANNEL_NAME = 'theia-electron-ipc';
+
+/**
+ * Electron-IPC-specific connection handler.
+ * Use this if you want to establish communication between the frontend and the electron-main process.
+ */
+export const ElectronConnectionHandler = Symbol('ElectronConnectionHandler');
```
Comment (akosyakov) [inline packages/core/src/electron-common/messaging/electron-connection-handler.ts:28]: I wonder whether we really need it. `ConnectionHandler` belongs to common module meaning that it can be used in any context already. It also does not imply any particular connection.

Is there an issue with using `ConnectionHandler` directly?

File: packages/core/src/node/cli.ts (hunk 6 above)
Comment (kittaakos) [reply packages/core/src/node/cli.ts:48]: Also see: https://github.com/eclipse-theia/theia/pull/7964#discussion_r439269907

File: packages/core/src/electron-browser/window/electron-window-service.ts [hunk 8]
```diff
@@ -14,20 +14,19 @@
  * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
  ********************************************************************************/
 
-import { injectable } from 'inversify';
-import { ipcRenderer } from 'electron';
+import { injectable, inject } from 'inversify';
 import { NewWindowOptions } from '../../browser/window/window-service';
 import { DefaultWindowService } from '../../browser/window/default-window-service';
+import { ElectronWindowService as ElectronMainWindowService } from '../../electron-common/electron-window-service';
```
Comment (akosyakov) [inline packages/core/src/electron-browser/window/electron-window-service.ts:20]: there are r files with `electron-window-service` name? https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#unique-names

Comment (kittaakos) [reply packages/core/src/electron-browser/window/electron-window-service.ts:20]: Good point, I have also noticed it. I will pick another name. Thanks!

File: packages/core/src/common/messaging/abstract-connection-provider.ts [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+     */
+    static createProxy<T extends object>(container: interfaces.Container, path: string, factory: JsonRpcProxyFactory<T>): JsonRpcProxy<T>;
+    /**
+     * Create a proxy object to remote interface of T type
+     * over an electron ipc connection for the given path.
+     *
+     * An optional target can be provided to handle
+     * notifications and requests from a remote side.
+     */
+    static createProxy<T extends object>(container: interfaces.Container, path: string, target?: object): JsonRpcProxy<T> {
```
Comment (akosyakov) [inline packages/core/src/common/messaging/abstract-connection-provider.ts:44]: I don't think we need code which does nothing at runtime. It should be rather removed and a proper method added to each provider.

File: examples/api-samples/package.json [hunk 10]
```diff
@@ -10,6 +10,12 @@
   "theiaExtensions": [
     {
       "frontend": "lib/browser/api-samples-frontend-module"
+    },
+    {
+      "electronMain": "lib/electron-main/update/sample-updater-main-module"
+    },
+    {
+      "frontendElectron": "lib/electron-browser/updater/sample-updater-frontend-module"
```
Comment (akosyakov) [inline examples/api-samples/package.json:18]: It should belong to the previous entry point?

File: packages/core/src/electron-main/electron-application.ts [hunk 11]
```diff
@@ -0,0 +1,440 @@
...
+    /**
+     * The application is stopping. Contributions must perform only synchronous operations.
+     */
+    onStop?(application?: ElectronApplication): void;
+}
+
+// Extracted the functionality from `yargs@15.4.0-beta.0`.
+// Based on https://github.com/yargs/yargs/blob/522b019c9a50924605986a1e6e0cb716d47bcbca/lib/process-argv.ts
+@injectable()
+export class ProcessArgv {
```
Comment (akosyakov) [inline packages/core/src/electron-main/electron-application.ts:84]: `ElectronProcessArgv`

File: packages/core/src/electron-main/electron-application.ts [hunk 12]
```diff
@@ -0,0 +1,440 @@
...
+
+    @inject(ProcessArgv)
+    protected processArgv: ProcessArgv;
+
+    protected readonly electronStore = new Storage();
+    protected readonly backendPort = new Deferred<number>();
+
+    protected config: FrontendApplicationConfig;
+
+    async start(config: FrontendApplicationConfig): Promise<void> {
```
Comment (akosyakov) [inline packages/core/src/electron-main/electron-application.ts:149]: Could you check in this class which methods should be exposed? I see not many are used outside on the app. What is the reasoning for others?

File: packages/core/src/electron-browser/window/electron-window-service.ts (hunk 8 above)
Comment (akosyakov) [reply packages/core/src/electron-browser/window/electron-window-service.ts:20]: btw `ElectronMainWindowService` sounds good

File: packages/core/src/node/cli.ts (hunk 6 above)
Comment (akosyakov) [reply packages/core/src/node/cli.ts:48]: I did not get the response there, on master we don't use construction injection. Using the construction injection is breaking API.

File: packages/core/src/electron-main/electron-application.ts (hunk 11 above)
Comment (kittaakos) [reply packages/core/src/electron-main/electron-application.ts:84]: I will change it. The idea was to have it here until we bump up the `yargs` version. Currently, it is in beta.

File: examples/api-samples/package.json (hunk 10 above)
Comment (kittaakos) [reply examples/api-samples/package.json:18]: Yes, true. Thanks!

**Comment by akosyakov** [comment]
My comments are minor, changes are looking good now. I have not tried though, waiting for the final version.

File: packages/core/src/electron-main/electron-application.ts [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+     * Start the NodeJS backend server.
+     *
+     * @return Running server's port promise.
+     */
+    protected async startBackend(): Promise<number> {
+        // Check if we should run everything as one process.
+        const noBackendFork = process.argv.indexOf('--no-cluster') !== -1;
+        // Any flag/argument passed after `--` will be forwarded to the backend process.
+        const backendArgvMarkerIndex = process.argv.indexOf('--');
+        const backendArgv = backendArgvMarkerIndex === -1 ? [] : process.argv.slice(backendArgvMarkerIndex + 1);
```
Comment (kittaakos) [inline packages/core/src/electron-main/electron-application.ts:352]: This is new. It took some time to figure out why the VS Code extensions do no start when I launch the app from the sources. Is there a chance to have it as before? This is a breaking change for all our clients for sure.

File: packages/core/src/electron-main/electron-application.ts (hunk 11 above)
Comment (akosyakov) [reply packages/core/src/electron-main/electron-application.ts:84]: You could try to bump it up in this PR. We don't need CQs for runtime deps. A license check from us is enough.

Comment (kittaakos) [reply packages/core/src/electron-main/electron-application.ts:84]: It's not easy. If I bump it up in Theia, there will be other (transitive) dependencies using an older version of `yargs`, and will mess up the hosting 😕

File: packages/core/src/electron-main/electron-application.ts (hunk 13 above)
Comment (marechal-p) [reply packages/core/src/electron-main/electron-application.ts:352]: The reasoning here is that app makers will most likely create their own cli, which ought to be more high level than what the backend cli expects. I think a breakage would be justified here, else we would have to defer the responsibility of re-routing command line arguments onto app makers.

Comment (kittaakos) [reply packages/core/src/electron-main/electron-application.ts:352]: >  I think a breakage would be justified here,

Yes, you're right but can we include it in a separate PR when it works? I do not understand why you want to put this CLI change to this PR. It was about DI and not changing the way we handle the CLI arguments.

> that app makers will most likely create their own cli

Do you have a real use-case where you want to use these new features? If yes, would you mind sharing them? If no, let's wait till we have a real need and do not make premature architectural decisions that require a breaking change.

Comment (marechal-p) [reply packages/core/src/electron-main/electron-application.ts:352]: > Yes, you're right but can we include it in a separate PR when it works?

I'm ok with this, let's do that.

> I do not understand why you want to put this CLI change to this PR. It was about DI and not changing the way we handle the CLI arguments.

While doing so I ended up breaking apart the electron-main thing we had before which was really hard to look at. Might have been carried away.

> Do you have a real use-case where you want to use these new features?

Yes, we are making an electron app at Ericsson and already worked around having DI to handle the app's CLI.

Am ok to do that later.

File: packages/core/src/electron-common/messaging/electron-connection-handler.ts (hunk 7 above)
Comment (marechal-p) [reply packages/core/src/electron-common/messaging/electron-connection-handler.ts:28]: It was just confused handling a connection handler in the electron main process. I knew these are used in the backend process, and seeing it again in the electron main process somehow gave me a hard time reasoning about what does what. Ultimately you are right, and this is how I used to have this implemented: using the `ConnectionHandler` symbol directly.

Also, I know it is a far stretch, but maybe later we'll have connection handlers for communication between backend and electron main process? In such a case, we cannot use the same `ConnectionHandler` symbol because it would be ambiguous as to which service we want to provide a connection handler to.

TL;DR: I found it confusing, but maybe the current naming isn't much more explanatory. Hopefully the jsdoc might help someone hovering the symbol.

File: packages/core/src/common/messaging/abstract-connection-provider.ts (hunk 9 above)
Comment (marechal-p) [reply packages/core/src/common/messaging/abstract-connection-provider.ts:44]: The class is abstract, and this is a workaround to having abstract static methods... Subclasses provide the actual implementation for what are helper functions anyway (doing `SomeConnectionProvider.createProxy(...)`).

This simplifies Inversify bindings, and respects the old API to not break clients.

Comment (marechal-p) [reply packages/core/src/common/messaging/abstract-connection-provider.ts:44]: Else we need to copy/paste the signature overloads + doc in each subclass.

File: packages/core/src/electron-common/messaging/electron-connection-handler.ts (hunk 7 above)
Comment (akosyakov) [reply packages/core/src/electron-common/messaging/electron-connection-handler.ts:28]: ok

File: packages/core/src/browser/messaging/ws-connection-provider.ts [hunk 14]
```diff
@@ -68,6 +68,24 @@ export class WebSocketConnectionProvider extends AbstractConnectionProvider<WebS
         }
     }
 
+    protected doOpenChannel(path: string, handler: (channel: WebSocketChannel) => void, options?: WebSocketOptions): void {
```
Comment (marechal-p) [inline packages/core/src/browser/messaging/ws-connection-provider.ts]: I'm sorry but I don't understand that change. How is it different from `super.openChannel(...)`?

File: packages/core/src/electron-main/electron-application.ts (hunk 13 above)
Comment (marechal-p) [reply packages/core/src/electron-main/electron-application.ts]: Reverted this change.

File: packages/core/src/node/cli.ts (hunk 6 above)
Comment (marechal-p) [reply packages/core/src/node/cli.ts]: Construction injection is used on this class on master. Although you are right and these changes are left overs.

File: packages/core/src/electron-main/electron-application.ts (hunk 12 above)
Comment (kittaakos) [reply packages/core/src/electron-main/electron-application.ts:150]: I left `start(config: FrontendApplicationConfig)`, `openNewWindow(url: string)`, `quit()`, and `get config()` as public and reduced the visibility of others.

File: packages/core/src/browser/messaging/ws-connection-provider.ts (hunk 14 above)
Comment (kittaakos) [reply packages/core/src/browser/messaging/ws-connection-provider.ts]: You're right! Thanks a lot 🙏

File: packages/core/src/common/messaging/abstract-connection-provider.ts (hunk 9 above)
Comment (kittaakos) [reply packages/core/src/common/messaging/abstract-connection-provider.ts:44]: I am fine with both. I leave as is.

**Comment by kittaakos** [comment]
Closing in favor of https://github.com/eclipse-theia/theia/pull/8125
